### DISCUSSION TITLE
Getting a weird linker error in my cpp code

### DISCUSSION THREAD

**Discussion by Aarnav-JP** [Homebrew/discussions#3659] [Everyday usage]
hi everyone , I am getting this error in my code I have never seen this before please help me resolve it - 

```
0  0x1003e81a0  __assert_rtn + 140
1  0x10026fa8c  mach_o::relocatable::Parser<arm64>::parse(mach_o::relocatable::ParserOptions const&) + 4536
2  0x100241d38  mach_o::relocatable::Parser<arm64>::parse(unsigned char const*, unsigned long long, char const*, long, ld::File::Ordinal, mach_o::relocatable::ParserOptions const&) + 148
3  0x1002aa4ac  ld::tool::InputFiles::makeFile(Options::FileInfo const&, bool) + 1468
4  0x1002ad360  ___ZN2ld4tool10InputFilesC2ER7Options_block_invoke + 56
5  0x1a67101f4  _dispatch_client_callout2 + 20
6  0x1a6724f8c  _dispatch_apply_invoke_and_wait + 224
7  0x1a672426c  _dispatch_apply_with_attr_f + 1152
8  0x1a672447c  dispatch_apply + 108
9  0x1002ad1f4  ld::tool::InputFiles::InputFiles(Options&) + 616
10  0x10022f6c0  main + 552
A linker snapshot was created at:
        /tmp/prac-2022-09-16-202054.ld-snapshot
ld: Assertion failed: (_file->_atomsArrayCount == computedAtomCount && "more atoms allocated than expected"), function parse, file macho_relocatable_file.cpp, line 2061.
collect2: error: ld returned 1 exit status
```

code - 

```
#include <bits/stdc++.h>

#define ll long long int
#define fori(a, b) for (ll i = a; i < b; i++)

#define all(x) (x).begin(), (x).end()

#define endl "\n"
#define fastinp                       \
    ios_base::sync_with_stdio(false); \
    cin.tie(NULL);

using namespace std;

void solve()
{
    ll n;
    cin >> n;
    vector<pair<ll, ll>> v(n);
    fori(0, n)
    {
        ll a;
        cin >> a;
        v.push_back({a, i + 1});
    }
    sort(all(v), greater<pair<ll, ll>>());
    ll sum = 0;
    ll cans = 0;
    fori(0, n)
    {
        sum += ((cans * v[i].first) + 1);
        cans++;
    }
    cout << sum << endl;
    for (auto it : v)
    {
        cout << it.second << " ";
    }
    cout << endl;
}

int main()
{
    fastinp;

    int t = 1;

    if (0)
    {
        cin >> t;
    }
    for (; t != 0; --t)
        solve();

    return 0;
}
```

**Comment by danjin250** [comment]
I don't think it's an issue with your code, I think there's just an actual linker bug.

You haven't mentioned your OS and XCode version but I'm guessing MacOS Monterey and XCode 13 or 14?
Perhaps you're suffering the same issue as these ones

https://github.com/iains/gcc-12-branch/issues/6
https://github.com/spack/spack/issues/32640

**Reply by Aarnav-JP** [reply]
thanks for your reply. yeah I am using Monterey and Xcode 14. So there is nothing that can be done from my side?

**Reply by teshields** [reply]
I am getting the same link error during portability tests of my application using the Homebrew GNU C++ compilers (version 11.3.0 & 12.2.0), after upgrading to Xcode 14.0 on macOS Monterey 12.6 (M1 chip). The application compiles & links & runs AOK with the Xcode 14 C++ compiler (Xcode is the development environment).

Probably unsurprisingly, the portability tests using the Homebrew LLVM C++ compilers (versions 11.1.0, 12.0.1, 13.0.1 & 14.0.6) see no problems.

**Reply by satyamk230ct** [reply]
Same error i'm getting after updating my macOs to Monterey 12.6 (M1 chip) and using Xcode 14.0

**Comment by satyamk230ct** [comment]
Please notify me if you find any solution for this

**Reply by amitsingh0823** [reply]
use gnu++14

**Comment by theharshpat** [comment]
Having the same issue after the Xcode update. macOS 12.5.1, Xcode Version 14.0 (14A309), M1 processor.

**Comment by Bo98** [comment]
Yes, this is a Xcode 14 bug. Apple are working on this. It should be fixed in a future Xcode beta (not 14.1 beta 1).

The only known workaround is downgrading to Xcode 13.4.

**Reply by theharshpat** [reply]
Thanks for the reply @Bo98 

I can confirm, that downgrading to 13.4 resolves this issue.

**Reply by abhipatel1331** [reply]
how to downgrad safely?

**Reply by theharshpat** [reply]
@abhipatel1331  You can follow the instructions, I have written a StackOverflow answer.

https://stackoverflow.com/a/73765819/7657049

Basically, download and install 13.4 command line tools, switch to that version using command and verify the version of Xcode command line tools.

**Reply by jaidetree** [reply]
Is there anyway to do that without an Apple Developer account? Not loving the idea of paying $99 just to download an older version of these tools.

**Reply by jaidetree** [reply]
Found a way! If you open xcode, Preferences > Accounts, and sign into your Apple Id. You'll have a free tier account.

**Reply by Bo98** [reply]
You can download with the free Apple Developer account.

**Reply by crisluengo** [reply]
Is it known if Xcode 14.1 beta 3 fixes this issue?

**Reply by InverseRE** [reply]
https://github.com/iains/gcc-12-branch/issues/6#issuecomment-1260282797

**Reply by NikithGanga** [reply]
Yes, it fixes the error, and it's very simple as well, search the beta version, login into ur Apple Developer Account and then download the latest beta of the corresponding Command Line Tools 14.1 beta 3 in this case, and it should fix the error after downloading the tools and, just Log Out, just in case to apply all the new changes

**Reply by jatin1510** [reply]
Please someone elaborate more. I can't fix this.

**Reply by NikithGanga** [reply]
Firstly, you have 2 choices : 
1)Either wait for the next update of Xcode, where the issue would be resolved 
or
2)Follow the below steps : 
- Please go to [https://developer.apple.com/download/all/](url) 
- Enter your Apple Credentials
- Now Scroll Down and Find the "Command Line Tools for Xcode 14.1 beta 3"
- Download the dmg file and install the package
- You can restart your Mac once, after the installation, for the changes to take effect
 - You will see that the issue is fixed

**Reply by LuisMiguelSS** [reply]
@NikithGanga your link seems to be pointing at a Github Homebrew discussion

Here's the link, for everyone's comfort: [https://developer.apple.com/download/all/](https://developer.apple.com/download/all/) ;)

**Reply by Techie232** [reply]
@NikithGanga  thank you very much brother!!

**Reply by jatin1510** [reply]
After upgrading to macOs sonoma, the issue remains. If we try to reinstall the old command line tools, package installer gives the error **macOs is too new.**

Any suggestions or quick fixes?

**Reply by NikithGanga** [reply]
@jatin1510 Relink or join the system link, which you intially used to compile your programs with.It should work now

**Reply by jatin1510** [reply]
Can not understand what you are saying...

**Comment by seye11** [comment]
I used -O2 to compile my cpp file, and it works!

- Xcode 14.0.1 on macOS Monterey 12.6(intel)

Hope it can help you.

**Comment by thinkyhead** [comment]
Under the MacPorts package system this did the trick for my purposes today (to be able to do a native build under PlatformIO):
```sh
sudo port install gcc12
cd /opt/local/bin
sudo rm -f gcc g++ cc
sudo ln -s gcc-mp-12 gcc ; sudo ln -s g++-mp-12 g++ ; sudo ln -s g++ cc
cd -
rehash
```
The same general idea should also work for Homebrew. We'll see if it continues to work under macOS 13 Ventura.

**Reply by Bo98** [reply]
Xcode 14.1 is released on Monday which will fix the issue. The RC is available now.

In fact if you're already using Ventura, I highly recommend using Xcode 14.1 RC. In Homebrew, we require it.

**Reply by thinkyhead** [reply]
In my case a separate gcc was needed. PlatformIO requires a real gcc and doesn't get along with Clang. But, is there some way to invoke the Xcode / macOS bundled gcc without invoking Clang?

**Reply by Bo98** [reply]
Oh, so this is different issue than the linker crash.

macOS & Xcode hasn't shipped with GCC for many years now (but still has a "gcc" executable pointing to clang), so yeah if you have something that absolutely needs GCC then you'll need to install GCC from MacPorts or Homebrew.

**Reply by chriselrod** [reply]
I get the same error with 14.2: https://github.com/JuliaSIMD/LoopModels/actions/runs/3777632683/jobs/6421565544

**Reply by egorpugin** [reply]
@chriselrod 
Look at line 335, you are using 14.0.1 xcode.
`-isysroot /Applications/Xcode_14.0.1.app`

Anyways github CI has 14.1 only at the moment.

**Reply by chriselrod** [reply]
Okay, I had an action that ran
```
> echo $(pkgutil --pkg-info=com.apple.pkg.CLTools_Executables)
package-id: com.apple.pkg.CLTools_Executables version: 14.2.0.0.1.1668646533 volume: / location: / install-time: 1671153287 groups: com.apple.FindSystemFiles.pkg-group
```
which is why I said 14.2.
But looking at the logs, you're obviously right.

**Reply by carlocab** [reply]
@chriselrod try
```
sudo xcode-select --switch /Library/Developer/CommandLineTools
```
to use the newer CLT SDK.

**Reply by chriselrod** [reply]
Great, that worked! No more linker error, but many tests segfault for some reason: https://github.com/JuliaSIMD/LoopModels/actions/runs/3784846071/jobs/6434459661
